### ISSUE TITLE
PyTorch 1.8.0 / 1.8.1 CUDA memory corruption on oddly sized tesnor sections of strided tensor

### ISSUE THREAD

**Issue by JuriAbramov** [pytorch/pytorch#56694]
## 🐛 Bug

PyTorch 1.8 introduced a CUDA memory corruption oddly sized tensor sections. This is a regression in 1.8, 1.7.1 did not have the issue and never did early 1.8 *alpha* versions.

## To Reproduce

0. start python with CUDA memcheck
1. make a 1x48x576x960 GPU tensor float16 "x" with strides like (26542080, 1, 46080, 48)
2. take f = [:,:23,...] section
3. make some ops like torch.sum(f, dim=1, keepdim=True)
4. check the crash and/or memory corruption.

## Expected behavior

This should just work. 

## Environment
python3 /h/roast/juri/collect_env.py
Collecting environment information...
PyTorch version: 1.8.1+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.4 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2

Python version: 3.6 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: Tesla P100-SXM2-16GB
GPU 1: Tesla P100-SXM2-16GB
GPU 2: Tesla P100-SXM2-16GB
GPU 3: Tesla P100-SXM2-16GB
GPU 4: Tesla P100-SXM2-16GB
GPU 5: Tesla P100-SXM2-16GB
GPU 6: Tesla P100-SXM2-16GB
GPU 7: Tesla P100-SXM2-16GB

Nvidia driver version: 455.23.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.18.2
[pip3] torch==1.8.1
[pip3] torch-tb-profiler==0.1.0
[conda] Could not collect

 - PyTorch Version (e.g., 1.0):1.8.1
 - OS (e.g., Linux): Lunux
 - How you installed PyTorch (`conda`, `pip`, source): pip, torch=1.8.1
 - Build command you used (if compiling from source):
 - Python version: 3.6

## Additional context
Obvious workaroumd is to use x.contiguous(), but finding such issue in complex environment is time consuming.

**Comment by ptrblck** [comment]
I cannot reproduce this issue on a P100 using the PyTorch 1.8.1 binaries with CUDA10.2 and `compute-sanitizer` doesn't show any errors.
Code:
```python
import torch

x = torch.randn(1, 48, 576, 960, dtype=torch.float16, device='cuda').to(memory_format=torch.channels_last)
print(x.shape)
print(x.stride())
f = x[:, :23, ...]
out = torch.sum(f, dim=1, keepdim=True)
print(out)
```
Output:
```python
torch.Size([1, 48, 576, 960])
(26542080, 1, 46080, 48)
tensor([[[[-2.0664,  2.2617, -3.7676,  ..., -5.1016,  3.1973, 11.0547],
          [-5.0469, -1.3145,  4.8086,  ...,  4.3477, -2.1035, -1.1309],
          [-1.0244, -3.1934, -0.9224,  ..., -3.2695,  2.0000,  2.9980],
          ...,
          [ 1.8584, -2.2363, -1.9121,  ..., -3.2773,  7.5000, -1.6475],
          [ 2.4902, -6.6914, -1.0039,  ..., -2.9961,  6.7539, -1.8652],
          [-8.3203, -3.2285, -3.9980,  ...,  2.8086,  1.7783,  4.4922]]]],
       device='cuda:0', dtype=torch.float16)
```
Env:
```
PyTorch version: 1.8.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.19.6

Python version: 3.6 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: Tesla P100-SXM2-16GB
GPU 1: Tesla P100-SXM2-16GB
GPU 2: Tesla P100-SXM2-16GB
GPU 3: Tesla P100-SXM2-16GB
GPU 4: Tesla P100-SXM2-16GB
GPU 5: Tesla P100-SXM2-16GB
GPU 6: Tesla P100-SXM2-16GB
GPU 7: Tesla P100-SXM2-16GB

Nvidia driver version: 450.51.06
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.0
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.8.1
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              hfd86e86_1
[conda] mkl                       2020.2                      256
[conda] mkl-service               2.3.0            py36he8ac12f_0
[conda] mkl_fft                   1.3.0            py36h54f3939_0
[conda] mkl_random                1.1.1            py36h0573a6f_0
[conda] numpy                     1.19.2           py36h54aff64_0
[conda] numpy-base                1.19.2           py36hfa32c7d_0
[conda] pytorch                   1.8.1           py3.6_cuda10.2_cudnn7.6.5_0    pytorch
```

Could you post an executable code snippet to reproduce this issue or do you see any issues in my code?

**Comment by JuriAbramov** [comment]
Unfortunately I cannot repro the issue in isolation either, and the whole model is proprietary and too complex.
Any idea is there are some attributes to the tensor beyond shape / stride / dtype / device?

In my case, I also see first a few hundred of issues in torch.cat like one below, and they are gone with with contiguous() call added.

========= CUDA-MEMCHECK
========= Invalid __global__ read of size 2
=========     at 0x00000b70 in void at::native::_GLOBAL__N__51_tmpxft_00002805_00000000_9_Shape_compute_70_cpp1_ii_cedd8df2::CatArrayBatchedCopy<c10::Half, unsigned int, int=4, int=64, int=64>(c10::Half*, at::n
ative::_GLOBAL__N__51_tmpxft_00002805_00000000_9_Shape_compute_70_cpp1_ii_cedd8df2::CatArrInputTensorMetadata<at::native::_GLOBAL__N__51_tmpxft_00002805_00000000_9_Shape_compute_70_cpp1_ii_cedd8df2::CatArrayBat
chedCopy<c10::Half, unsigned int, int=4, int=64, int=64>, unsigned int, int=64, int=64>, at::native::_GLOBAL__N__51_tmpxft_00002805_00000000_9_Shape_compute_70_cpp1_ii_cedd8df2::TensorSizeStride<at::native::_GL
OBAL__N__51_tmpxft_00002805_00000000_9_Shape_compute_70_cpp1_ii_cedd8df2::CatArrInputTensorMetadata, unsigned int=4>, int, at::native::_GLOBAL__N__51_tmpxft_00002805_00000000_9_Shape_compute_70_cpp1_ii_cedd8df2
::CatArrInputTensorMetadata)
=========     by thread (63,0,0) in block (81,0,0)
=========     Address 0x7efaefead880 is out of bounds
=========     Saved host backtrace up to driver entry point at kernel launch time
=========     Host Frame:/usr/lib/x86_64-linux-gnu/libcuda.so.1 (cuLaunchKernel + 0x2b8) [0x1e5cc8]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libcudart-80664282.so.10.2 [0x1a547]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libcudart-80664282.so.10.2 [0x1a5d7]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libcudart-80664282.so.10.2 (cudaLaunchKernel + 0x225) [0x51885]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libtorch_cuda.so [0x27fc9d4]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libtorch_cuda.so (_ZN2at6native75_GLOBAL__N__51_tmpxft_00002805_00000000_9_Shape_compute_70_cpp1_ii_cedd8df212parallel_catIN3c104Hal
fELi64ELi64EEEvRNS_6TensorERKNS3_8ArrayRefIS5_EEliNS3_12MemoryFormatE + 0x2f8) [0x28107f8]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libtorch_cuda.so (_ZN2at6native12cat_out_cudaERNS_6TensorEN3c108ArrayRefIS1_EEl + 0x1741) [0x27feaa1]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libtorch_cuda.so (_ZN2at6native8cat_cudaEN3c108ArrayRefINS_6TensorEEEl + 0x8c) [0x27ff28c]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libtorch_cuda.so [0xea154b]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libtorch_cuda.so [0xea1598]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so [0x1608aa1]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so (_ZN2at4_catEN3c108ArrayRefINS_6TensorEEEl + 0x65) [0x148f3c5]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so (_ZN2at6native3catEN3c108ArrayRefINS_6TensorEEEl + 0xf7) [0x11ad8d7]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so [0x1750ab8]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so [0x1750b38]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so [0x1608aa1]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so (_ZN2at3catEN3c108ArrayRefINS_6TensorEEEl + 0x65) [0x148ee25]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so [0x2cb7487]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so [0x2cb7a88]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so [0x1608aa1]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so (_ZN2at3catEN3c108ArrayRefINS_6TensorEEEl + 0x65) [0x148ee25]
=========     Host Frame:/u/juri/.local/lib/python3.6/site-packages/torch/lib/libtorch_python.so [0x4a1c08]
